### PR TITLE
add rpc call to resend missing cross shard receipt

### PR DESCRIPTION
**Pull request by chaosma**

### CODE REVIEW

File: internal/hmyapi/backend.go [hunk 1]
```diff
@@ -64,6 +64,9 @@ type Backend interface {
 	// Get balance
 	GetBalance(address common.Address) (*hexutil.Big, error)
 	GetShardID() uint32
+
+	// retrieve the blockHash using txID and add blockHash to CxPool for resending
+	AddBlockHashToCxPool(ctx context.Context, txID common.Hash) (uint64, bool)
```
Comment (harmony-ek) [inline internal/hmyapi/backend.go]: Name interface methods after what they achieve, not after how they are implemented:

```suggestion
	ResendCx(ctx context.Context, txID common.Hash) (uint64, bool)
```

File: node/node_cross_shard.go [hunk 2]
```diff
@@ -64,6 +64,33 @@ func (node *Node) BroadcastCXReceipts(newBlock *types.Block, lastCommits []byte)
 	}
 }
 
+// BroadcastMissingCXReceipts broadcasts missing cross shard receipts per request
+func (node *Node) BroadcastMissingCXReceipts() {
+	sendNextTime := []common.Hash{}
+	it := node.CxPool.Pool().Iterator()
+	for blockHash := range it.C {
+		blk := node.Blockchain().GetBlockByHash(blockHash.(common.Hash))
```
Comment (harmony-ek) [inline node/node_cross_shard.go]: Please do not use `GetBlockByHash` and later `GetBlockByNumber` just to use its header.  Use `GetHeaderByHash` and `GetHeaderByNumber` instead.

File: node/node_cross_shard.go [hunk 3]
```diff
@@ -64,6 +64,33 @@ func (node *Node) BroadcastCXReceipts(newBlock *types.Block, lastCommits []byte)
...
+		blockNum := blk.NumberU64()
+		nextBlk := node.Blockchain().GetBlockByNumber(blockNum + 1)
+		if nextBlk == nil {
+			sendNextTime = append(sendNextTime, blockHash.(common.Hash))
+			continue
+		}
+		sig := nextBlk.Header().LastCommitSignature()
+		bitmap := nextBlk.Header().LastCommitBitmap()
+		lastCommits := append(sig[:], bitmap...)
+		go node.BroadcastCXReceipts(blk, lastCommits)
```
Comment (harmony-ek) [inline node/node_cross_shard.go]: Doesn't this broadcast to all N shards?  Can we make it broadcast to just the shard in which the tx is found?

(In order for this to work, you might have to track `(blockHash, toShard)` tuple in the pool.)

File: internal/hmyapi/blockchain.go [hunk 4]
```diff
@@ -112,6 +112,12 @@ func (s *PublicBlockChainAPI) BlockNumber() hexutil.Uint64 {
 	return hexutil.Uint64(header.Number().Uint64())
 }
 
+// ResendTx retrieves blockHash by TxID and add blockHash to CxPool for resending
```
Comment (harmony-ek) [inline internal/hmyapi/blockchain.go]:
```suggestion
// ResendCx requests that the egress receipt for the given cross-shard
// transaction be sent to the destination shard for credit.  This is used for
// unblocking a half-complete cross-shard transaction whose fund has been
// withdrawn already from the source shard but not credited yet in the
// destination account due to transient failures.
```

File: internal/hmyapi/blockchain.go [hunk 5]
```diff
@@ -112,6 +112,12 @@ func (s *PublicBlockChainAPI) BlockNumber() hexutil.Uint64 {
...
+func (s *PublicBlockChainAPI) ResendTx(ctx context.Context, txID common.Hash) (bool, error) {
```
Comment (harmony-ek) [inline internal/hmyapi/blockchain.go]:
```suggestion
func (s *PublicBlockChainAPI) ResendCx(ctx context.Context, txID common.Hash) (bool, error) {
```

Comment (harmony-ek) [reply internal/hmyapi/blockchain.go]: By the way, is resending egress CX receipts a responsibility of a blockchain API proper?  Egress CX sending is an implementation detail of a cross-shard transaction, which is at the protocol level…  Maybe we need to revisit elsewhere to root this.

File: core/blockchain.go [hunk 6]
```diff
@@ -2237,3 +2237,9 @@ func (bc *BlockChain) UpdateCXReceiptsCheckpointsByBlock(block *types.Block) {
 		bc.updateCXReceiptsCheckpoints(k, v)
 	}
 }
+
+// ReadTxLookupEntry returns the corresponding blockHash and blockNum where transaction locates
+func (bc *BlockChain) ReadTxLookupEntry(txID common.Hash) (common.Hash, uint64) {
+	blockHash, blockNum, _ := rawdb.ReadTxLookupEntry(bc.db, txID)
```
[the remark this answers is not in the document]
Comment (chaosma) [reply core/blockchain.go]: expose the index

[the remark this answers is not in the document]
Comment (chaosma) [reply core/blockchain.go]: the cache is in the low level function already

File: hmy/backend.go [hunk 7]
```diff
@@ -75,6 +77,9 @@ func New(nodeAPI NodeAPI, txPool *core.TxPool, eventMux *event.TypeMux, shardID
 // TxPool ...
 func (s *Harmony) TxPool() *core.TxPool { return s.txPool }
 
+// CxPool ...
+func (s *Harmony) CxPool() *core.CxPool { return s.cxPool }
```
[the remark this answers is not in the document]
Comment (chaosma) [reply hmy/backend.go]: add comments

File: node/node_cross_shard.go (hunk 2 above)
Comment (chaosma) [reply node/node_cross_shard.go]: good point. changed

File: internal/hmyapi/blockchain.go (hunk 4 above)
Comment (chaosma) [reply internal/hmyapi/blockchain.go]: Thanks for detailed description

File: node/node_cross_shard.go (hunk 3 above)
Comment (chaosma) [reply node/node_cross_shard.go]: changed the data structure to be (blockHash, toShardID); also added tx type checking

File: core/cx_pool.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+// CxEntry represents the egress receipt's blockHash and ToShardID
+type CxEntry struct {
+	BlockHash common.Hash
+	ToShardID uint32
+}
+
+// CxPool is to hold a pool of block outgoing receipts to be resend in next round broadcast
+type CxPool struct {
+	pool    mapset.Set
```
Comment (rlan35) [inline core/cx_pool.go]: Add comment for what's the expected data to be stored in this set.

File: node/node_cross_shard.go [hunk 9]
```diff
@@ -64,6 +64,71 @@ func (node *Node) BroadcastCXReceipts(newBlock *types.Block, lastCommits []byte)
...
+
+	groupID := p2p.ShardID(toShardID)
+	go node.host.SendMessageToGroups([]p2p.GroupID{p2p.NewGroupIDByShardID(groupID)}, host.ConstructP2pMessage(byte(0), proto_node.ConstructCXReceiptsProof(cxReceipts, merkleProof, newBlock.Header(), commitSig, commitBitmap)))
+}
+
+// BroadcastMissingCXReceipts broadcasts missing cross shard receipts per request
+func (node *Node) BroadcastMissingCXReceipts() {
+	sendNextTime := []core.CxEntry{}
+	it := node.CxPool.Pool().Iterator()
+	for entry := range it.C {
```
Comment (rlan35) [inline node/node_cross_shard.go]: Will this loop just exit once there is no item to iterate?

File: core/cx_pool.go (hunk 8 above)
Comment (chaosma) [reply core/cx_pool.go]: added

**Pull request merged**